### PR TITLE
feat: add uglify token to interpolateName

### LINKED ISSUE #174: RFC: add uglify token to interpolateName
### Summary

moved from https://github.com/webpack-contrib/file-loader/issues/380
How about add uglify token to `interpolateName`?
If we use ugly name in production, we can get lighter bundle file.

### Syntax

- `[ugly-name:<uglyType>]`
  - uglyType => `word`, `file`, `css`, etc... (default: word)

Reference: https://github.com/TroyTae/uglify-set

### PR DESCRIPTION
**Pull request by TroyTae**
resolve #174

### CODE REVIEW

File: lib/interpolateName.js [hunk 1]
```diff
@@ -3,11 +3,16 @@
...
 const getHashDigest = require('./getHashDigest');
+const { WordUglify, FileUglify, CssUglify } = require('uglify-set');
 
 const emojiRegex = /[\uD800-\uDFFF]./;
 const emojiList = emojisList.filter((emoji) => emojiRegex.test(emoji));
 const emojiCache = {};
 
+const wordUglify = WordUglify();
+const fileUglify = FileUglify();
+const cssUglify = CssUglify();
```
Comment (evilebottnawi) [inline lib/interpolateName.js:14]: Potential memory leak in watch mode

Comment (TroyTae) [reply lib/interpolateName.js:14]: Oh you're right
Is there a way to detect it?

Comment (TroyTae) [reply lib/interpolateName.js:14]: It seems memory leak can fire anywhere not only here.
To solve this I'm adding forcing reset code in `uglify-set`.
But there is no specification maximum size of string in Ecma.
We need some standard, if you have some idea plz tell me :)
I'll also research about this.

Comment (TroyTae) [reply lib/interpolateName.js:14]: http://www.ecma-international.org/ecma-262/6.0/#sec-array-exotic-objects
According to spec, `length` of Array is less than 2^32.
How about follow this?

Comment (evilebottnawi) [reply lib/interpolateName.js:14]: only weak cache can solve this problem

Comment (TroyTae) [reply lib/interpolateName.js:14]: Maybe I'm not catch the point now.
Can you explain it more detail?

File: lib/interpolateName.js [hunk 2]
```diff
@@ -3,11 +3,26 @@
 const path = require('path');
 const emojisList = require('emojis-list');
 const getHashDigest = require('./getHashDigest');
+const {
+  DigitUglify,
+  AlphabetUglify,
+  WordUglify,
+  FileUglify,
+  CssUglify,
+} = require('uglify-set');
```
Comment (TroyTae) [inline lib/interpolateName.js:12]: continuously https://github.com/webpack/loader-utils/pull/175#discussion_r476575919 by @evilebottnawi

Comment (TroyTae) [inline lib/interpolateName.js:12]: continuously https://github.com/webpack/loader-utils/pull/175#discussion_r476575919 by @evilebottnawi

File: lib/interpolateName.js [hunk 3]
```diff
@@ -3,11 +3,26 @@
...
 const emojiList = emojisList.filter((emoji) => emojiRegex.test(emoji));
 const emojiCache = {};
 
+const uglify = {
+  digit: DigitUglify(),
+  alphabet: AlphabetUglify(),
+  word: WordUglify(),
+  file: FileUglify(),
+  css: CssUglify(),
+};
```
Comment (TroyTae) [inline lib/interpolateName.js:24]: continuously https://github.com/webpack/loader-utils/pull/175#discussion_r476627900 by @evilebottnawi

Comment (evilebottnawi) [reply lib/interpolateName.js:24]: Still memory leak

Comment (TroyTae) [reply lib/interpolateName.js:24]: Yes this changes aren't for fixing memory leak.
I told like this in the previous comment:
> Maybe I'm not catch the point now.
> Can you explain it more detail?
Can you tell cause of memory leak or test method?

Comment (evilebottnawi) [reply lib/interpolateName.js:24]: @TroyTae just run your projects on very very very big code base, you can even write test in own repository

Comment (TroyTae) [reply lib/interpolateName.js:24]: image
(Capture snapshot after executing 9 million times)
Hmmm.. Can you tell me how do you test it?
It seems garbage collection works well...

Comment (evilebottnawi) [reply lib/interpolateName.js:24]: You need to run your tool in watch mode

Comment (TroyTae) [reply lib/interpolateName.js:24]: Above capture is tested in watch mode.
I made a webapp and run it using `webpack-dev-server`.

Comment (TroyTae) [reply lib/interpolateName.js:24]: I added force reset code when over 2 ** 16.
And this is the test code:
https://github.com/TroyTae/uglify-set/blob/master/__tests__/test-util.js#L13-L25
Can you check this again?

Comment (TroyTae) [reply lib/interpolateName.js:24]: @evilebottnawi Anything is blocking this?

**Comment by TroyTae** [comment]
@evilebottnawi 
Can you check this please?

**Comment by TroyTae** [comment]
@sokra 
It seems evilebottnawi is busy.
So if you are ok, can you check this instead?
I really wanna apply this feature on production before release.

**Pull request closed without merging**